controllers/argocd/notifications/util.go

```diff
@@ -0,0 +1,664 @@
+  oncePer: app.status.operationState.syncResult.revision
+  send:
+  - app-deployed
+  when: app.status.operationState.phase in ['Succeeded'] and app.status.health.status
+      == 'Healthy'`
+
+	notificationsConfig["trigger.on-health-degraded"] = `- description: Application has degraded
+  send:
+  - app-health-degraded
+  when: app.status.health.status == 'Degraded'`
+
+	notificationsConfig["trigger.on-sync-failed"] = `- description: Application syncing has failed
+  send:
+  - app-sync-failed
+  when: app.status.operationState.phase in ['Error', 'Failed']`
+
+	notificationsConfig["trigger.on-sync-running"] = `- description: Application is being synced
+  send:
+  - app-sync-running
+  when: app.status.operationState.phase in ['Running']`
+
+	notificationsConfig["trigger.on-sync-status-unknown"] = `- description: Application status is 'Unknown'
+  send:
+  - app-sync-status-unknown
+  when: app.status.sync.status == 'Unknown'`
+
+	notificationsConfig["trigger.on-sync-succeeded"] = `- description: Application syncing has succeeded
+  send:
+  - app-sync-succeeded
+  when: app.status.operationState.phase in ['Succeeded']`
+
+	return notificationsConfig
+}
+
+func GetNotificationsCommand(cr *argoprojv1a1.ArgoCD) []string {
+
+	cmd := make([]string, 0)
+	cmd = append(cmd, "argocd-notifications")
+
+	cmd = append(cmd, "--loglevel")
+	cmd = append(cmd, argoutil.GetLogLevel(cr.Spec.Notifications.LogLevel))
+
+	return cmd
+}
+
+// GetNotificationsResources will return the ResourceRequirements for the Notifications container.
+func GetNotificationsResources(cr *argoprojv1a1.ArgoCD) corev1.ResourceRequirements {
+	resources := corev1.ResourceRequirements{}
+
+	// Allow override of resource requirements from CR
+	if cr.Spec.Notifications.Resources != nil {
+		resources = *cr.Spec.Notifications.Resources
+	}
+
+	return resources
+}
+
+func GetArgoCDNotificationsControllerReplicas(nr *NotificationsReconciler) *int32 {
+	if nr.Instance.Spec.Notifications.Replicas != nil && *nr.Instance.Spec.Notifications.Replicas >= 0 {
+		return nr.Instance.Spec.Notifications.Replicas
+	}
+
+	return nil
+}
+
+func UpdateIfChanged(existingVal, desiredVal interface{}, extraAction func(), changed *bool) {
+	if !reflect.DeepEqual(existingVal, desiredVal) {
+		reflect.ValueOf(existingVal).Elem().Set(reflect.ValueOf(desiredVal).Elem())
+		if extraAction != nil {
+			extraAction()
+		}
+		*changed = true
+	}
+}
+
+// TODO: The below functions should be moved to some sharing util package.
+func GetArgoContainerImage(cr *argoprojv1a1.ArgoCD) string {
+	defaultTag, defaultImg := false, false
+	img := cr.Spec.Image
+	if img == "" {
+		img = common.ArgoCDDefaultArgoImage
+		defaultImg = true
+	}
+
+	tag := cr.Spec.Version
+	if tag == "" {
+		tag = common.ArgoCDDefaultArgoVersion
+		defaultTag = true
+	}
+	if e := os.Getenv(common.ArgoCDImageEnvVar); e != "" && (defaultTag && defaultImg) {
+		return e
+	}
+
+	return argoutil.CombineImageTag(img, tag)
+}
```

Comment:
This function belongs in a common package accessible to all controller packages 

I will move it in a separate PR